libkbfs/bserver_tlf_journal.go

```diff
@@ -0,0 +1,706 @@
+	[]byte, BlockCryptKeyServerHalf, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	return s.getDataLocked(id, context)
+}
+
+func (s *bserverTlfJournal) getAll() (
+	map[BlockID]map[BlockRefNonce]blockRefLocalStatus, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	if s.isShutdown {
+		return nil, errBserverTlfJournalShutdown
+	}
+
+	res := make(map[BlockID]map[BlockRefNonce]blockRefLocalStatus)
+
+	for id, refs := range s.refs {
+		if len(refs) == 0 {
+			continue
+		}
+
+		res[id] = make(map[BlockRefNonce]blockRefLocalStatus)
+		for ref, refEntry := range refs {
+			res[id][ref] = refEntry.Status
+		}
+	}
+
+	return res, nil
+}
+
+func (s *bserverTlfJournal) putData(
+	id BlockID, context BlockContext, buf []byte,
+	serverHalf BlockCryptKeyServerHalf) error {
+	err := validateBlockServerPut(s.crypto, id, context, buf)
+	if err != nil {
+		return err
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.isShutdown {
+		return errBserverTlfJournalShutdown
+	}
+
+	_, existingServerHalf, err := s.getDataLocked(id, context)
+	var exists bool
+	switch err.(type) {
+	case BServerErrorBlockNonExistent:
+		exists = false
+	case nil:
+		exists = true
+	default:
+		return err
+	}
+
+	if exists {
+		// If the entry already exists, everything should be
+		// the same, except for possibly additional
+		// references.
+
+		// We checked that both buf and existingData hash to
+		// id, so no need to check that they're both equal.
+
+		if existingServerHalf != serverHalf {
+			return fmt.Errorf(
+				"key server half mismatch: expected %s, got %s",
+				existingServerHalf, serverHalf)
+		}
+	}
+
+	err = os.MkdirAll(s.blockPath(id), 0700)
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(s.blockDataPath(id), buf, 0600)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Add integrity-checking for key server half?
+
+	err = ioutil.WriteFile(
+		s.keyServerHalfPath(id), serverHalf.data[:], 0600)
+	if err != nil {
+		return err
+	}
+
+	err = s.putRefEntryLocked(id, blockRefEntry{
+		Status:  liveBlockRef,
+		Context: context,
+	})
+	if err != nil {
+		return err
+	}
+
+	return s.appendJournalEntryLocked("put", id, []BlockContext{context})
+}
+
+func (s *bserverTlfJournal) addReference(id BlockID, context BlockContext) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.isShutdown {
+		return errBserverTlfJournalShutdown
+	}
+
+	refs := s.refs[id]
+	if refs == nil {
+		return BServerErrorBlockNonExistent{fmt.Sprintf("Block ID %s "+
+			"doesn't exist and cannot be referenced.", id)}
+	}
+
+	// Only add it if there's a non-archived reference.
+	hasNonArchivedRef := false
+	for _, refEntry := range refs {
+		if refEntry.Status == liveBlockRef {
+			hasNonArchivedRef = true
+			break
+		}
+	}
+
+	if !hasNonArchivedRef {
+		return BServerErrorBlockArchived{fmt.Sprintf("Block ID %s has "+
+			"been archived and cannot be referenced.", id)}
+	}
+
+	// TODO: Allow adding a reference even if all the existing
```

Comment:
From the perspective of the journal, I'm not sure this is a valid thing to do.  In fact, the journal might only want to allow reference to block IDs for which it has the corresponding block saved locally to disk.  When we use this as a real journal, we'll have to deal with cases where the block itself gets deleted from the bserver while an "addReference" call is still pending in a journal somewhere.  If the block contents isn't saved alongside that "addReference", I think we'll have no way to recover the data intended to be put by the user.

In fact, we'll have the difficult task of either re-putting the same block with the same ID, which may not currently be supported by the actual bserver, or fixing up the reference with a new block ID in the corresponding directory entry, which will also be very hard.  I think we talked about similar situations when writing up the design, but I can't remember if we covered this exact one or not.